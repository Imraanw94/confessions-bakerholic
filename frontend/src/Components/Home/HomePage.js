import { Card, Container, Stack } from "react-bootstrap";
import { Instagram } from "react-bootstrap-icons";

// This is the default page when a user visits the site. It'll just display some images of products, and an Instagram link
function HomePage(props) {
  // const oreoBall = require("./../../resources/images/oreo ball.webp");
  // const allChoc = require("./../../resources/images/all choc cookies.webp");

  return (
    <Container style={{ marginTop: "15px" }} className="home">
      <Card body className="text-center">
        <h1 style={{ fontStyle: "italic", fontSize: "52px" }}>
          Welcome to Confessions of a Bakerholic
        </h1>
      </Card>
      <Stack direction="horizontal" className="mb-3 image-stack" gap={2}>
        {/* <Stack direction="vertical">
          <Card.Img
            src={oreoBall}
            style={{ maxHeight: "500px" }}
          />
        </Stack>
        <Stack direction="vertical">
          <Card.Img
            src={allChoc}
            style={{ maxHeight: "500px" }}
          />
        </Stack> */}
      </Stack>
      <Card>
        <Card.Body className="text-center">
          <Card.Title>
            <h2>Follow Us On Instagram!</h2>
          </Card.Title>
          <Instagram
            className="me-1"
            style={{ fontSize: "32px", marginTop: "-12px" }}
          />
          <Card.Link
            style={{ fontSize: "32px" }}
            href="https://www.instagram.com/confessionsofabakerholic/"
            target="_blank"
          >
            Instagram Link
          </Card.Link>
        </Card.Body>
      </Card>
    </Container>
  );
}

export default HomePage;
