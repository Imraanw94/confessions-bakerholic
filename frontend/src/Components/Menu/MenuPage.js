import { useEffect, useState } from "react";
import { Card, Container, Row, Stack } from "react-bootstrap";
import { getCookie } from "../../cookieService";
import AddProduct from "./Products/AddProduct";
import Product from "./Products/Product";
import AddRecommendedProduct from "./Recommended/AddRecommendedProduct";
import RecommendedProduct from "./Recommended/RecommendedProduct";

// This page is where users can view products for sale by the business
// Or they can recommend products to sell
function MenuPage(props) {
  /**
   * @type {array}
   * productsData is used to store the data received from the API call
   */
  const [productsData, setProductsData] = useState([]);
  /**
   * @type {array}
   * recommendedData is used to store the data received from the API call
   */
  const [recommendedData, setRecommendedData] = useState([]);
  /**
   * @type {boolean}
   * tokenExists is used to indicate whether there is a token or not
   */
  const [tokenExists, setTokenExists] = useState(false);
  /**
   * @type {string}
   * token is used to store the token cookie
   */
  const [token, setToken] = useState(getCookie("token"));
  /**
   * @type {boolean}
   * tokenExists is used to indicate whether the user is an admin or not
   */
  const [isAdmin, setIsAdmin] = useState(false);

  // UserId used to perform CRUD operations
  const userId = getCookie("user");

  // This function gets the cookies of admin and token and sets the boolean value of each
  function checkCookie() {
    let isAdminCheck = getCookie("admin");
    setIsAdmin(isAdminCheck === "true");
    let tokenCheck = getCookie("token");
    setTokenExists(tokenCheck !== "");
  }

  // This function is used to fetch the Products in the DB
  async function fetchProducts() {
    const response = await fetch("/allproducts", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
    })
      .then((response) => {
        if (response.ok) {
          return response.json();
        }
        throw response;
      })
      .then((data) => {
        setProductsData(data);
      })
      .catch((error) => {
        console.error("Error fetching data: ", error);
      });
  }

  // This function is used to fetch the Recommended Products in the API
  // We use a POST instead of a GET because we have to pass the user ID to fetch the relevant Product Recommendations
  async function fetchRecommendedProducts() {
    const response = await fetch("/allrecommended", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify({ userId: userId }),
    })
      .then((response) => {
        if (response.ok) {
          return response.json();
        }
        throw response;
      })
      .then((data) => {
        setRecommendedData(data);
      })
      .catch((error) => {
        console.error("Error fetching data: ", error);
      });
  }

  // Asynchronous function for adding a Product
  async function addProductHandler(product) {
    const response = await fetch("/addproduct", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify(product),
    })
      .then((response) => {
        if (response.ok) {
          return response.json();
        }
        throw response;
      })
      .then((data) => {
        setProductsData(data);
        alert("Added information of a new product");
      });
  }

  // Asynchronous function for editing a Product
  async function editProductHandler(product) {
    const response = await fetch("/editproduct", {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify(product),
    })
      .then((response) => {
        if (response.ok) {
          return response.json();
        }
        throw response;
      })
      .then((data) => {
        setProductsData(data);
      });
  }

  // Asynchronous function for deleting a Product
  async function deleteProductHandler(id) {
    const response = await fetch("/deleteproduct", {
      method: "DELETE",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify({ productId: id.value, userId: userId }),
    })
      .then((response) => {
        if (response.ok) {
          return response.json();
        }
        throw response;
      })
      .then((data) => {
        setProductsData(data);
      });
  }

  // Asynchronous function for adding a Product Recommendation
  async function addRecommendationHandler(recommended) {
    const response = await fetch("/addrecommended", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify(recommended),
    })
      .then((response) => {
        if (response.ok) {
          return response.json();
        }
        throw response;
      })
      .then((data) => {
        setRecommendedData(data);
      });
  }

  // Asynchronous function for editing a Product Recommendation
  async function editRecommendationHandler(recommended) {
    const response = await fetch("/editrecommended", {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify(recommended),
    })
      .then((response) => {
        if (response.ok) {
          return response.json();
        }
        throw response;
      })
      .then((data) => {
        setRecommendedData(data);
      });
  }

  // Asynchronous function for deleting a Product Recommendation
  async function deleteRecommendationHandler(id) {
    const response = await fetch("/deleterecommended", {
      method: "DELETE",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify({ recommendedId: id.value, userId: userId }),
    })
      .then((response) => {
        if (response.ok) {
          return response.json();
        }
        throw response;
      })
      .then((data) => {
        setRecommendedData(data);
      });
  }

  // Since this is a functional component, we can use the useEffect hook to
  // fetch the data.
  // Empty array indicates to call the fetch function on initial load of the
  // page
  // Reviewers often tell me to check for warnings, but this is the cause of those warnings
  // This is intended functionality
  useEffect(() => {
    fetchProducts();
    checkCookie(); // Check cookies on initial load
    if (tokenExists) {
      fetchRecommendedProducts();
    }
  }, []);

  return (
    <Container style={{ marginTop: "15px" }}>
      <Card body className="mb-3 menu-header header text-center">
        <h1 style={{ fontStyle: "italic" }}>
          These are some of the delicious baked goods we're selling
        </h1>
      </Card>

      {isAdmin && (
        <Row>
          <AddProduct onAddProduct={addProductHandler}></AddProduct>
        </Row>
      )}

      <hr />
      <Row>
        <Stack direction="horizontal" gap={2} className="products grid-stack">
          {productsData.map((product, index) => (
            <Product
              key={product._id}
              id={product._id}
              name={product.name}
              description={product.description}
              price={product.price}
              userId={userId}
              isAdmin={isAdmin}
              onEditProduct={editProductHandler}
              onDeleteProduct={deleteProductHandler}
            />
          ))}
        </Stack>
      </Row>

      <hr />
      {tokenExists && (
        <AddRecommendedProduct
          onAddRecommendedProduct={addRecommendationHandler}
        />
      )}

      <hr />
      <Row>
        <Stack
          direction="horizontal"
          gap={2}
          className="recommended grid-stack"
        >
          {recommendedData.map((recommended, index) => (
            <RecommendedProduct
              key={recommended._id}
              id={recommended._id}
              name={recommended.name}
              recipe={recommended.recipe}
              picture={recommended.picture}
              userId={userId}
              isAdmin={isAdmin}
              onEditRecommendedProduct={editRecommendationHandler}
              onDeleteRecommendedProduct={deleteRecommendationHandler}
            />
          ))}
        </Stack>
      </Row>
    </Container>
  );
}

export default MenuPage;
