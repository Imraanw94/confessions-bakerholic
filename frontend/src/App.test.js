import renderer from "react-test-renderer";
import AboutPage from "./Components/About/AboutPage";
import ContactPage from "./Components/Contact/ContactPage";
import HomePage from './Components/Home/HomePage';
import MenuPage from "./Components/Menu/MenuPage";


test("snapshot test of HomePage", () => {
  const tree = renderer.create(<HomePage />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("snapshot test of AboutPage", () => {
  const tree = renderer.create(<AboutPage />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("snapshot test of MenuPage", () => {
  const tree = renderer.create(<MenuPage />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("snapshot test of ContactPage", () => {
  const tree = renderer.create(<ContactPage />).toJSON();
  expect(tree).toMatchSnapshot();
});