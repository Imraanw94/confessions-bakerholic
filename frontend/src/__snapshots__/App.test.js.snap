// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test of AboutPage 1`] = `
<div
  className="container"
  style={
    Object {
      "marginTop": "15px",
    }
  }
>
  <div
    className="vstack gap-3"
  >
    <div
      className="card"
    >
      <div
        className="card-body"
      >
        <div
          className="card-title h5"
        >
          <h3>
            About
          </h3>
        </div>
        <p
          className="card-text"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Fusce et placerat eros. Nullam sem eros, dapibus eu efficitur quis, tincidunt vitae lorem. Cras vitae viverra tellus. Pellentesque faucibus risus ac semper auctor. Aenean dictum mauris ornare vestibulum suscipit. Fusce sagittis leo scelerisque consectetur pellentesque. Integer nec ante maximus, viverra sapien eget, pulvinar urna. Suspendisse ut volutpat leo, vitae tempus est. Mauris ullamcorper, velit vel dapibus molestie, erat turpis sodales turpis, lacinia dignissim tellus nisi sit amet dui. Sed placerat ligula non orci porttitor vehicula.
        </p>
      </div>
    </div>
    <div
      className="card"
    >
      <div
        className="card-body"
      >
        <div
          className="card-title h5"
        >
          <h3>
            Location
          </h3>
        </div>
        <p
          className="card-text"
        >
          Vivamus a augue luctus, feugiat massa vitae, condimentum sapien. Nunc condimentum non sapien non efficitur. Pellentesque at orci a mi varius varius et non ex. Vivamus consectetur sem pellentesque nisi lobortis elementum. Donec et tellus at ligula suscipit lacinia at a mi. Curabitur ipsum felis, luctus eu bibendum nec, consectetur cursus ipsum. Donec tempor purus a ante mollis, eu rhoncus nisl tincidunt. Mauris sem mauris, porta vitae pharetra sed, iaculis eget felis. Aliquam vitae urna eu nunc finibus aliquet pretium in risus. Integer ligula metus, sagittis eget vulputate ac, hendrerit at purus. Fusce lectus sem, luctus placerat libero nec, dapibus accumsan dui. Nulla a luctus purus.
        </p>
      </div>
    </div>
    <div
      className="card"
    >
      <div
        className="card-body"
      >
        <div
          className="card-title h5"
        >
          <h3>
            Delivery
          </h3>
        </div>
        <p
          className="card-text"
        >
          Donec consectetur bibendum porttitor. Sed mi metus, porta eu diam a, blandit malesuada sem. Ut ut ex eu enim maximus tempor. Maecenas nec gravida nulla, vitae luctus enim. Aenean sit amet luctus massa. In sit amet tempus enim, ut egestas dolor. Curabitur maximus placerat mollis. Proin finibus ultricies nunc eu vestibulum.
        </p>
      </div>
    </div>
    <div
      className="card"
    >
      <div
        className="card-body"
      >
        <div
          className="card-title h5"
        >
          <h3>
            FAQ
          </h3>
        </div>
        <p
          className="card-text"
        >
          Cras non pulvinar erat, quis blandit sapien. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Maecenas pretium porta dapibus. Sed imperdiet sollicitudin velit, in cursus nunc. Fusce sed lorem porta, aliquet ante sit amet, pulvinar ipsum. Nam augue quam, consequat id erat sed, mollis pulvinar risus. Morbi nulla turpis, gravida vel euismod nec, rhoncus ac mi. Duis sem dui, malesuada sit amet lectus eget, ornare suscipit mauris. Suspendisse potenti.
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`snapshot test of ContactPage 1`] = `
<div
  className="contact container"
  style={
    Object {
      "marginTop": "15px",
    }
  }
>
  <div
    className="mb-3 card"
  >
    <div
      className="card-body"
    >
      <h1>
        Contact Us
      </h1>
    </div>
  </div>
  <div
    className="mb-3 hstack"
    style={
      Object {
        "justifyContent": "center",
      }
    }
  >
    <div
      className="me-3 card"
    >
      <div
        className="card-body"
      >
        <div
          className="card-title h5"
        >
          Name
        </div>
        <div
          className="mb-3"
        >
          <input
            className="form-control"
            id="formBasicName"
            placeholder="Your Name"
            type="text"
          />
        </div>
      </div>
    </div>
    <div
      className="me-3 card"
    >
      <div
        className="card-body"
      >
        <div
          className="card-title h5"
        >
          Surname
        </div>
        <div
          className="mb-3"
        >
          <input
            className="form-control"
            id="formBasicSurname"
            placeholder="Your Surname"
            type="text"
          />
        </div>
      </div>
    </div>
    <div
      className="me-3 card"
    >
      <div
        className="card-body"
      >
        <div
          className="card-title h5"
        >
          Email
        </div>
        <div
          className="mb-3"
        >
          <input
            className="form-control"
            id="formBasicEmail"
            placeholder="Your Email"
            type="text"
          />
        </div>
      </div>
    </div>
    <div
      className="me-3 card"
    >
      <div
        className="card-body"
      >
        <div
          className="card-title h5"
        >
          Phone
        </div>
        <div
          className="mb-3"
        >
          <input
            className="form-control"
            id="formBasicPhone"
            placeholder="Your Phone Number"
            type="text"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="p-3 card"
  >
    <form
      className=""
    >
      <div
        className="mb-3"
      >
        <label
          className="form-label"
          htmlFor="formBasicAdditionalInfo"
        >
          Additional Info
        </label>
        <textarea
          className="form-control"
          id="formBasicAdditionalInfo"
          placeholder="Additonal Information"
          rows={3}
        />
      </div>
      <button
        className="btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;

exports[`snapshot test of HomePage 1`] = `
<div
  className="home container"
  style={
    Object {
      "marginTop": "15px",
    }
  }
>
  <div
    className="text-center card"
  >
    <div
      className="card-body"
    >
      <h1
        style={
          Object {
            "fontSize": "52px",
            "fontStyle": "italic",
          }
        }
      >
        Welcome to Confessions of a Bakerholic
      </h1>
    </div>
  </div>
  <div
    className="mb-3 image-stack hstack gap-2"
  >
    <div
      className="vstack"
    >
      <img
        className="card-img"
        src="oreo ball.webp"
        style={
          Object {
            "maxHeight": "500px",
          }
        }
      />
    </div>
    <div
      className="vstack"
    >
      <img
        className="card-img"
        src="all choc cookies.webp"
        style={
          Object {
            "maxHeight": "500px",
          }
        }
      />
    </div>
  </div>
  <div
    className="card"
  >
    <div
      className="text-center card-body"
    >
      <div
        className="card-title h5"
      >
        <h2>
          Follow Us On Instagram!
        </h2>
      </div>
      <svg
        className="me-1"
        fill="currentColor"
        height="1em"
        style={
          Object {
            "fontSize": "32px",
            "marginTop": "-12px",
          }
        }
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 0C5.829 0 5.556.01 4.703.048 3.85.088 3.269.222 2.76.42a3.917 3.917 0 0 0-1.417.923A3.927 3.927 0 0 0 .42 2.76C.222 3.268.087 3.85.048 4.7.01 5.555 0 5.827 0 8.001c0 2.172.01 2.444.048 3.297.04.852.174 1.433.372 1.942.205.526.478.972.923 1.417.444.445.89.719 1.416.923.51.198 1.09.333 1.942.372C5.555 15.99 5.827 16 8 16s2.444-.01 3.298-.048c.851-.04 1.434-.174 1.943-.372a3.916 3.916 0 0 0 1.416-.923c.445-.445.718-.891.923-1.417.197-.509.332-1.09.372-1.942C15.99 10.445 16 10.173 16 8s-.01-2.445-.048-3.299c-.04-.851-.175-1.433-.372-1.941a3.926 3.926 0 0 0-.923-1.417A3.911 3.911 0 0 0 13.24.42c-.51-.198-1.092-.333-1.943-.372C10.443.01 10.172 0 7.998 0h.003zm-.717 1.442h.718c2.136 0 2.389.007 3.232.046.78.035 1.204.166 1.486.275.373.145.64.319.92.599.28.28.453.546.598.92.11.281.24.705.275 1.485.039.843.047 1.096.047 3.231s-.008 2.389-.047 3.232c-.035.78-.166 1.203-.275 1.485a2.47 2.47 0 0 1-.599.919c-.28.28-.546.453-.92.598-.28.11-.704.24-1.485.276-.843.038-1.096.047-3.232.047s-2.39-.009-3.233-.047c-.78-.036-1.203-.166-1.485-.276a2.478 2.478 0 0 1-.92-.598 2.48 2.48 0 0 1-.6-.92c-.109-.281-.24-.705-.275-1.485-.038-.843-.046-1.096-.046-3.233 0-2.136.008-2.388.046-3.231.036-.78.166-1.204.276-1.486.145-.373.319-.64.599-.92.28-.28.546-.453.92-.598.282-.11.705-.24 1.485-.276.738-.034 1.024-.044 2.515-.045v.002zm4.988 1.328a.96.96 0 1 0 0 1.92.96.96 0 0 0 0-1.92zm-4.27 1.122a4.109 4.109 0 1 0 0 8.217 4.109 4.109 0 0 0 0-8.217zm0 1.441a2.667 2.667 0 1 1 0 5.334 2.667 2.667 0 0 1 0-5.334z"
        />
      </svg>
      <a
        className="card-link"
        href="https://www.instagram.com/confessionsofabakerholic/"
        style={
          Object {
            "fontSize": "32px",
          }
        }
        target="_blank"
      >
        Instagram Link
      </a>
    </div>
  </div>
</div>
`;

exports[`snapshot test of MenuPage 1`] = `
<div
  className="container"
  style={
    Object {
      "marginTop": "15px",
    }
  }
>
  <div
    className="mb-3 menu-header header text-center card"
  >
    <div
      className="card-body"
    >
      <h1
        style={
          Object {
            "fontStyle": "italic",
          }
        }
      >
        These are some of the delicious baked goods we're selling
      </h1>
    </div>
  </div>
  <hr />
  <div
    className="row"
  >
    <div
      className="products grid-stack hstack gap-2"
    />
  </div>
  <hr />
  <hr />
  <div
    className="row"
  >
    <div
      className="recommended grid-stack hstack gap-2"
    />
  </div>
</div>
`;
